In [1]:
import pandas as pd 
import numpy as np
df1=pd.read_csv('Datasets/tmdb_5000_credits.csv');
df2=pd.read_csv('Datasets/tmdb_5000_movies.csv');

### Joining the 2 datasets about the column 'id'


In [2]:
df1.columns=['id','title','cas','crew']
df2=df2.merge(df1,on='id')

In [3]:
df2.head(3)

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  300000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2  245000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                       homepage      id  \
0                   http://www.avatarmovie.com/   19995   
1  http://disney.go.com/disneypictures/pirates/     285   
2   http://www.sonypictures.com/movies/spectre/  206647   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...                en   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   

                                            overview  popularity  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577   
1  Captain Barbossa, long believed to be dead, ha...  139.082615   
2  A cryptic message from Bond’s past sends him o...  107.376788   

                                production_companies  ... runtime  \
0  [{"name": "Ingenious Film Partners", "id": 289...  ...   162.0   
1  [{"name": "Walt Disney Pictures", "id": 2}, {"...  ...   169.0   
2  [{"name": "Columbia Pictures", "id": 5}, {"nam...  ...   148.0   

                                    spoken_languages    status  \
0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
1           [{"iso_639_1": "en", "name": "English"}]  Released   
2  [{"iso_639_1": "fr", "name": "Fran\u00e7ais"},...  Released   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   
2                           A Plan No One Escapes   

                                    title_x vote_average vote_count  \
0                                    Avatar          7.2      11800   
1  Pirates of the Caribbean: At World's End          6.9       4500   
2                                   Spectre          6.3       4466   

                                    title_y  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   

                                                 cas  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  

[3 rows x 23 columns]

# Demographic Filtering 
    -We can use the average ratings of the movie as a score to measure the likeness or to rate a movie.
    DISADVANTAGE: It wont be fair because a movie with a rating of 9.2 average rating and only 4 votes cannot just be considered better than a movie with 7.8 as average but 40 votes.
     

## IMDb's weighted rating 
 -Hence i chose to use the following weighted rating criteria(wr) given as:-
     Weighted Rating(WR)=((v/(v+m))*R)+((m/(v+m))*C)
     where,
     -v is the number of votes for the movie.
     -m is the number of votes required to be listed in the chart.
     -R is the average rating of the movie.
     -C is the mean vote across the whole report.
     
     

In [4]:
C=df2['vote_average'].mean(); # mean rating for all m
C

6.092171559442011

In [5]:
# Using 90th percentile as cutoff.
m=df2['vote_count'].quantile(0.9)
m

1838.4000000000015

In [6]:
q_movies=df2.copy().loc[df2['vote_count']>=m]
q_movies.shape

(481, 23)

#### This tells us that only 481 movies are there which qualify to be in this list. 

## Defining a function to calculate the weighted rating of movies 

In [11]:
def weighted_rating(x,m=m,C=C):
    v=x['vote_count']
    R=x['vote_average']
    # Applying the formula
    
    return (v/(v+m)*R)+(m/(m+v)*C)
    

In [12]:
#defining a new feature 'score' and calculating its value with ' weighted rating function'

q_movies['score']= q_movies.apply(weighted_rating,axis=1)

In [13]:
# Sorting the movies based on the score obtained 
q_movies = q_movies.sort_values('score', ascending=False)

In [36]:
#Printing the top 15 movies
q_movies.head(15)

budget                                             genres  \
1881   25000000  [{"id": 18, "name": "Drama"}, {"id": 80, "name...   
662    63000000                      [{"id": 18, "name": "Drama"}]   
65    185000000  [{"id": 18, "name": "Drama"}, {"id": 28, "name...   
3232    8000000  [{"id": 53, "name": "Thriller"}, {"id": 80, "n...   
96    160000000  [{"id": 28, "name": "Action"}, {"id": 53, "nam...   
3337    6000000  [{"id": 18, "name": "Drama"}, {"id": 80, "name...   
95    165000000  [{"id": 12, "name": "Adventure"}, {"id": 18, "...   
809    55000000  [{"id": 35, "name": "Comedy"}, {"id": 18, "nam...   
329    94000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
1990   18000000  [{"id": 12, "name": "Adventure"}, {"id": 28, "...   
262    93000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2912   11000000  [{"id": 12, "name": "Adventure"}, {"id": 28, "...   
1818   22000000  [{"id": 18, "name": "Drama"}, {"id": 36, "name...   
3865    3300000                      [{"id": 18, "name": "Drama"}]   
330    79000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   

                                               homepage      id  \
1881                                                NaN     278   
662          http://www.foxmovies.com/movies/fight-club     550   
65         http://thedarkknight.warnerbros.com/dvdsite/     155   
3232                                                NaN     680   
96                http://inceptionmovie.warnerbros.com/   27205   
3337                       http://www.thegodfather.com/     238   
95                    http://www.interstellarmovie.net/  157336   
809                                                 NaN      13   
329                       http://www.lordoftherings.net     122   
1990  http://www.starwars.com/films/star-wars-episod...    1891   
262                      http://www.lordoftherings.net/     120   
2912  http://www.starwars.com/films/star-wars-episod...      11   
1818                     http://www.schindlerslist.com/     424   
3865                  http://sonyclassics.com/whiplash/  244786   
330                      http://www.lordoftherings.net/     121   

                                               keywords original_language  \
1881  [{"id": 378, "name": "prison"}, {"id": 417, "n...                en   
662   [{"id": 825, "name": "support group"}, {"id": ...                en   
65    [{"id": 849, "name": "dc comics"}, {"id": 853,...                en   
3232  [{"id": 396, "name": "transporter"}, {"id": 14...                en   
96    [{"id": 1014, "name": "loss of lover"}, {"id":...                en   
3337  [{"id": 131, "name": "italy"}, {"id": 699, "na...                en   
95    [{"id": 83, "name": "saving the world"}, {"id"...                en   
809   [{"id": 422, "name": "vietnam veteran"}, {"id"...                en   
329   [{"id": 603, "name": "elves"}, {"id": 606, "na...                en   
1990  [{"id": 526, "name": "rebel"}, {"id": 803, "na...                en   
262   [{"id": 603, "name": "elves"}, {"id": 604, "na...                en   
2912  [{"id": 803, "name": "android"}, {"id": 4270, ...                en   
1818  [{"id": 1382, "name": "factory"}, {"id": 1631,...                en   
3865  [{"id": 1416, "name": "jazz"}, {"id": 1523, "n...                en   
330   [{"id": 603, "name": "elves"}, {"id": 606, "na...                en   

                                         original_title  \
1881                           The Shawshank Redemption   
662                                          Fight Club   
65                                      The Dark Knight   
3232                                       Pulp Fiction   
96                                            Inception   
3337                                      The Godfather   
95                                         Interstellar   
809                                        Forrest Gump   
329       The Lord of the Rings: The Return of the 

In [33]:
q_movies.rename(columns={'title_x':'Title'},inplace=True)

### Obtaining the first 15 most  

In [43]:
ans=q_movies[['Title','vote_count','score']].head(15)
ans.style.hide_index()
